test/irmin-pack/test_inode.ml

```diff
@@ -664,10 +666,145 @@ module Inode_tezos = struct
+    (module T.Inter.Child_ordering)
+
+  let check_child_index pos (module Order : S) ~reference ~step ~depth =
+    let msg =
+      Fmt.str "Short hash of child at { depth = %d; step = %S }" depth step
+    in
+    let expected = reference ~depth step in
+    let actual = Order.key step |> Order.index ~depth in
+    check_int pos ~msg ~expected actual
+
+  let check_max_depth_exception pos (module Order : S) ~step ~depth =
+    match Order.key step |> Order.index ~depth with
+    | index ->
+        Alcotest.failf ~pos
+          "Expected [Max_depth %d] to be raised, but got a computed index of \
+           %d instead"
+          depth index
+    | exception Irmin_pack.Inode.Max_depth _ -> ()
+
+  (* Get the bit at index [n] in a string: *)
+  let get_bit str n =
+    let chosen_byte = Bytes.get_uint8 (Bytes.unsafe_of_string str) (n / 8) in
+    let bit_index_in_byte = n mod 8 in
+    (* Selects only the chosen bit from our byte: *)
+    let mask = 1 lsl (7 - bit_index_in_byte) in
+    let masked_byte = chosen_byte land mask in
+    let chosen_bit = masked_byte lsr (7 - bit_index_in_byte) in
+    assert (chosen_bit = 0 || chosen_bit = 1);
+    chosen_bit
+
+  let test_seeded_hash () =
+    let entries = Irmin_tezos.Conf.entries in
+    let reference ~depth step =
+      abs (Step.short_hash ~seed:depth step) mod entries
+    in
+    let (module Order) = make `Seeded_hash in
+    for _ = 1 to 1_000 do
+      let step = random_string 8 and depth = Random.int 10 in
+      check_child_index __POS__ (module Order) ~reference ~step ~depth
+    done
+
+  let hash_bits_max_depth ~log2_entries =
+    (* For a given [depth], the final bit of the corresponding index is at
+     * position [log2_entries * depth + log2_entries - 1] in the hash. If this
+     * is out-of-bounds in the hash, then we expect computing the ordering to
+     * fail (since we don't use modular indexing of the hash). *)
+    let rec aux depth =
+      if log2_entries * (depth + 1) > 8 * Hash.hash_size then depth - 1
+      else aux (succ depth)
+    in
+    aux 0
+
+  let test_hash_bits () =
+    (* [entries] is required to be a power of 2 greater than 1 and less than
+       2048, so we test every possible value here: *)
+    for log2_entries = 1 to 10 do
+      let entries = 1 lsl log2_entries in
+      let max_depth = hash_bits_max_depth ~log2_entries in
+      let (module Order) = make ~entries `Hash_bits in
+      [%log.app
+        "Testing hash_bits with { log_entries = %d; entries = %d; max_depth = \
+         %d }"
+        log2_entries entries max_depth];
+
+      (* Index is computed by reading [log2_entries] consecutive bits from the
+         hash of the step, starting at the [log2_entries * depth]-th byte. *)
+      let reference ~depth step =
+        let hash = Step.hash step in
+        let index = ref 0 in
+        for i = 0 to log2_entries - 1 do
+          let selected_bit = get_bit hash ((log2_entries * depth) + i) in
```

Comment:
I'm a bit confused, shouldn't `selected_bit` be `selected_bits` ? Since `index` returns `[log2_entries]` bits not just one?

Reply:
Each call to `get_bit` picks only a single bit from the hash, but this line is inside a `for` loop that runs `log2_entries`-many times (shifting a new bit onto the end of `index` each time), so I think "`selected_bit`" is still correct.

Reply:
ah yes, thanks